File: src/lib/anp/signing.ts
/**
 * Ed25519 identity and the three signed strings defined by ANP/0.1:
 * the registration proof (section 2), the canonical request string
 * (section 3), and the event authorship string (section 4).
 *
 * Uses `globalThis.crypto.subtle` only, so it runs unmodified in Node 18+
 * and in browsers that implement WebCrypto Ed25519.
 */

import {
  fromBase64Url,
  sha256Hex,
  toBase64Url,
  utf8,
} from "./encoding";

/** The protocol version string baked into every signature and hash. */
export const PROTOCOL_VERSION = "ANP/0.1";

const ED25519 = { name: "Ed25519" } as const;

/**
 * An agent identity: a raw 32-byte Ed25519 keypair. The private key never
 * needs to leave the process that generated it; only `publicKey` (base64url)
 * is ever sent to a host. `agentId` is assigned by the host at registration
 * and filled in by {@link register}.
 */
export interface AgentIdentity {
  /** Raw 32-byte Ed25519 public key, base64url encoded. */
  publicKey: string;
  /** SHA-256 hex of the raw public key bytes, the human-facing identity. */
  fingerprint: string;
  /** WebCrypto handle for signing. Not extractable by default. */
  privateKey: CryptoKey;
  /** WebCrypto handle for local verification. */
  publicCryptoKey: CryptoKey;
  /** Host-assigned registration id (UUID), set after registration. */
  agentId?: string;
}

/**
 * Detect WebCrypto Ed25519 support in the current runtime. Some browsers
 * (and older Node versions) do not implement the algorithm; callers should
 * check this once and present a clear message instead of a thrown error.
 */
export async function ed25519Supported(): Promise<boolean> {
  try {
    const pair = (await globalThis.crypto.subtle.generateKey(ED25519, false, [
      "sign",
      "verify",
    ])) as CryptoKeyPair;
    return Boolean(pair.privateKey && pair.publicKey);
  } catch {
    return false;
  }
}

/**
 * Generate a fresh Ed25519 identity. The private key is created
 * non-extractable: it can sign, but its bytes cannot be exported.
 */
export async function generateIdentity(): Promise<AgentIdentity> {
  // extractable=false applies to the private key; public keys are always
  // exportable, which is all we need for the raw key and its fingerprint.
  const pair = (await globalThis.crypto.subtle.generateKey(ED25519, false, [
    "sign",
    "verify",
  ])) as CryptoKeyPair;
  const raw = new Uint8Array(
    await globalThis.crypto.subtle.exportKey("raw", pair.publicKey),
  );
  return {
    publicKey: toBase64Url(raw),
    fingerprint: await sha256Hex(raw),
    privateKey: pair.privateKey,
    publicCryptoKey: pair.publicKey,
  };
}

/** Import a base64url raw Ed25519 public key for verification. */
export async function importPublicKey(
  publicKeyBase64Url: string,
): Promise<CryptoKey> {
  return globalThis.crypto.subtle.importKey(
    "raw",
    fromBase64Url(publicKeyBase64Url) as BufferSource,
    ED25519,
    true,
    ["verify"],
  );
}

/** Fingerprint of a base64url public key: SHA-256 hex of the raw bytes. */
export async function fingerprintOf(
  publicKeyBase64Url: string,
): Promise<string> {
  return sha256Hex(fromBase64Url(publicKeyBase64Url));
}

/** Sign a UTF-8 string (or raw bytes) and return the base64url signature. */
export async function sign(
  privateKey: CryptoKey,
  message: string | Uint8Array,
): Promise<string> {
  const bytes = typeof message === "string" ? utf8(message) : message;
  const sig = await globalThis.crypto.subtle.sign(
    ED25519,
    privateKey,
    bytes as BufferSource,
  );
  return toBase64Url(new Uint8Array(sig));
}

/** Verify a base64url signature over a UTF-8 string or raw bytes. */
export async function verifySignature(
  publicKeyBase64Url: string,
  signatureBase64Url: string,
  message: string | Uint8Array,
): Promise<boolean> {
  try {
    const key = await importPublicKey(publicKeyBase64Url);
    const bytes = typeof message === "string" ? utf8(message) : message;
    return await globalThis.crypto.subtle.verify(
      ED25519,
      key,
      fromBase64Url(signatureBase64Url) as BufferSource,
      bytes as BufferSource,
    );
  } catch {
    return false;
  }
}

/**
 * The registration proof string of section 2. Signing it proves possession
 * of the private key matching `publicKeyBase64Url`.
 *
 * `ANP/0.1\nregister\n<public_key_base64url>`
 */
export function registrationProofString(publicKeyBase64Url: string): string {
  return `${PROTOCOL_VERSION}\nregister\n${publicKeyBase64Url}`;
}

/**
 * The canonical request string of section 3, signed into the
 * `x-anp-signature` header of every authenticated request.
 *
 * `ANP/0.1\n<METHOD>\n<PATH>\n<timestamp>\n<nonce>\n<sha256_hex_of_body>`
 *
 * @param method uppercase HTTP method
 * @param path URL path only, no query string, no host
 * @param bodySha256Hex SHA-256 hex of the exact request body bytes; for
 * bodyless requests, the hash of the empty string
 */
export function canonicalRequestString(
  method: string,
  path: string,
  timestamp: string,
  nonce: string,
  bodySha256Hex: string,
): string {
  return `${PROTOCOL_VERSION}\n${method}\n${path}\n${timestamp}\n${nonce}\n${bodySha256Hex}`;
}

/**
 * The event authorship string of section 4, signed by the party submitting
 * an event: `ANP/0.1\n<kind>\n<payload_hash>`.
 */
export function authorshipString(kind: string, payloadHash: string): string {
  return `${PROTOCOL_VERSION}\n${kind}\n${payloadHash}`;
}

File: packages/anp-client-py/anp_client/signing.py
"""Ed25519 identity and the signed strings of ANP/0.1 sections 2 to 4."""

from __future__ import annotations

import base64
import hashlib
import secrets
from dataclasses import dataclass, field

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

PROTOCOL_VERSION = "ANP/0.1"

EMPTY_BODY_SHA256 = (
    "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
)
"""SHA-256 hex of the empty string, the body hash for bodyless requests."""


def to_base64url(data: bytes) -> str:
    """Base64url without padding, per RFC 4648 section 5."""
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def from_base64url(text: str) -> bytes:
    return base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))


def sha256_hex(data: str | bytes) -> str:
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def random_nonce() -> str:
    """A fresh request nonce, well under the 120 character limit."""
    return to_base64url(secrets.token_bytes(18))


@dataclass
class AgentIdentity:
    """A raw 32-byte Ed25519 keypair. Only ``public_key`` ever leaves the
    process; ``agent_id`` is assigned by the host at registration."""

    private_key: Ed25519PrivateKey
    public_key: str
    fingerprint: str
    agent_id: str | None = field(default=None)


def generate_identity() -> AgentIdentity:
    """Generate a fresh Ed25519 identity."""
    private_key = Ed25519PrivateKey.generate()
    raw = private_key.public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw
    )
    return AgentIdentity(
        private_key=private_key,
        public_key=to_base64url(raw),
        fingerprint=sha256_hex(raw),
    )


def fingerprint_of(public_key_base64url: str) -> str:
    """Fingerprint of a base64url raw public key: SHA-256 hex of the bytes."""
    return sha256_hex(from_base64url(public_key_base64url))


def sign(identity: AgentIdentity, message: str | bytes) -> str:
    """Sign a UTF-8 string or raw bytes; returns the base64url signature."""
    if isinstance(message, str):
        message = message.encode("utf-8")
    return to_base64url(identity.private_key.sign(message))


def verify_signature(
    public_key_base64url: str, signature_base64url: str, message: str | bytes
) -> bool:
    """Verify a base64url Ed25519 signature. Never raises; returns False."""
    if isinstance(message, str):
        message = message.encode("utf-8")
    try:
        key = Ed25519PublicKey.from_public_bytes(
            from_base64url(public_key_base64url)
        )
        key.verify(from_base64url(signature_base64url), message)
        return True
    except (InvalidSignature, ValueError):
        return False


def registration_proof_string(public_key_base64url: str) -> str:
    """Section 2: ``ANP/0.1\\nregister\\n<public_key_base64url>``."""
    return f"{PROTOCOL_VERSION}\nregister\n{public_key_base64url}"


def canonical_request_string(
    method: str, path: str, timestamp: str, nonce: str, body_sha256_hex: str
) -> str:
    """Section 3 canonical request string; ``path`` is the URL path only."""
    return f"{PROTOCOL_VERSION}\n{method}\n{path}\n{timestamp}\n{nonce}\n{body_sha256_hex}"


def authorship_string(kind: str, payload_hash: str) -> str:
    """Section 4: ``ANP/0.1\\n<kind>\\n<payload_hash>``."""
    return f"{PROTOCOL_VERSION}\n{kind}\n{payload_hash}"
